audition/distance_from_best.py

```diff
@@ -0,0 +1,300 @@
+                row should be a dict with keys:
+                        'metric' (e.g. 'precision@')
+                        'param' (e.g. '100_abs')
+                All models should have the results.evaluations table populated
+                for all given model group ids, train end times, and metric/param combos
+        """
+        for metric in metrics:
+            self.db_engine.execute('''
+                insert into {new_table}
+                WITH model_ranks AS (
+                    SELECT
+                        m.model_group_id,
+                        m.model_id,
+                        m.train_end_time,
+                        ev.value,
+                        row_number() OVER (
+                            PARTITION BY m.train_end_time
+                            ORDER BY ev.value DESC, RANDOM()
+                        ) AS rank
+                  FROM results.evaluations ev
+                  JOIN results.{models_table} m USING(model_id)
+                  JOIN results.model_groups mg USING(model_group_id)
+                  WHERE ev.metric='{metric}' AND ev.parameter='{metric_param}'
+                        AND m.model_group_id IN ({model_group_ids})
+                        AND train_end_time in ({train_end_times})
+                ),
+                model_tols AS (
+                  SELECT train_end_time, model_group_id, model_id,
+                         rank,
+                         value,
+                         first_value(value) over (
+                            partition by train_end_time
+                            order by rank ASC
+                        ) AS best_val
+                  FROM model_ranks
+                ),
+                current_best_vals as (
+                    SELECT
+                        model_group_id,
+                        model_id,
+                        train_end_time,
+                        '{metric}',
+                        '{metric_param}',
+                        value as raw_value,
+                        best_val - value below_best
+                    FROM model_tols
+                )
+                select
+                    current_best_vals.*,
+                    first_value(below_best) over (
+                        partition by model_group_id
+                        order by train_end_time asc
+                        rows between 1 following and unbounded following
+                    ) below_best_next_time
+                from current_best_vals
+            '''.format(
+                model_group_ids=str_in_sql(model_group_ids),
+                train_end_times=str_in_sql(train_end_times),
+                models_table=self.models_table,
+                metric=metric['metric'],
+                metric_param=metric['param'],
+                new_table=self.distance_table
+            ))
+
+    def create_and_populate(
+        self,
+        model_group_ids,
+        train_end_times,
```

Comment:
I'm not entirely certain why, but I found that I had to sort the train_end_times otherwise the points in the best distance graph were out of order. That doesn't make a ton of sense, but probably a good practice to sort these when they come in regardless.